src/main.rs

```diff
@@ -0,0 +1,116 @@
+use std::env;
+use uuid::Uuid;
+use openssl::error::ErrorStack;
+use openssl::ssl::{SslConnector, SslMethod};
+use postgres::{error::SqlState, Client, Error, Transaction};
+use postgres_openssl::MakeTlsConnector;
+
+// BEGIN ssl_config
+fn ssl_config() -> Result<MakeTlsConnector, ErrorStack> {
+    let builder = SslConnector::builder(SslMethod::tls())?;
+    Ok(MakeTlsConnector::new(builder.build()))
+}
+// END ssl_config
+
+/// Runs op inside a transaction and retries it as needed.
+/// On non-retryable failures, the transaction is aborted and
+/// rolled back; on success, the transaction is committed.
+// BEGIN execute_txn
+fn execute_txn<T, F>(client: &mut Client, op: F) -> Result<T, Error>
+where
+    F: Fn(&mut Transaction) -> Result<T, Error>,
+{
+    let mut txn = client.transaction()?;
+    loop {
+        let mut sp = txn.savepoint("cockroach_restart")?;
+        match op(&mut sp).and_then(|t| sp.commit().map(|_| t)) {
```

Comment:
I'm finding it a little tough to figure out what this logic is doing. It might be clearer to write it a bit more imperatively, without `and_then` or `unwrap_or`.

Reply:
I agree, though I'm not comfortable enough in Rust to refactor this. Another question for @ericharmeling.